### DISCUSSION TITLE
more realistic sonar data

### DISCUSSION THREAD

**Discussion by JustLuckyRobotics** [cyberbotics/wrestling#52] [Ideas]
I've been doing some testing with the sonar sensors on the body of the robot and i think I've discovered potential for improvement. 

(from nao.proto line 762)
type "sonar"
numberOfRays 6
aperture 1.04

The current implementation is set up to use 6 rays which look to be on the circumference of a given cone which is does work and give distances, however the data is inconsistent and doesn't behave as a sonar sensor should/would work. as there is very few rays to calculate the distance, you can have the enemy directly in front of the sensor and it it does not trigger as it is between these rays, this would not happen with a true sonar.

After testing modifying the the number of rays in the nao.proto file, it can approximate the data a sonar would give more closely, I haven't detected any noticeable performance drops when running with >60 rays but not sure of the ideal number.

Would it be possible to modify the proto to accommodate this change?

**Comment by omichel** [comment]
Sure. However I would appreciate if you could hint a suggestion on what could be a reasonable number of rays.

Increasing from 6 to 60 may be overkill and does impact the simulation speed in fast mode. On my computer, the performance drops from 4x to 3.2x (as displayed in the speedometer) if I set the number of rays to 60 on each sonar of each robot and enable them with a 20 ms time step. If set the number of rays to 30, the performance is 3.6x. Given that this concerns only one CPU core and won't affect the amount of CPU power allocated to the controllers (because they run on different cores of the CPU), it is acceptable to impact the performance a little bit, as long as it doesn't fall below 1.0x. So, I would appreciate if you could provide a good compromise value because the figures provided here may be different on other computers and I would like to keep a security margin with the performance of the simulation.

**Comment by omichel** [comment]
@JustLuckyRobotics: do you have any hint on the ideal number of rays?

**Reply by JustLuckyRobotics** [reply]
sorry i forgot to reply. i have done a fair amount of testing and i think 18 would be a good compromise. more does improve the data but diminishing returns start about 18.

**Reply by omichel** [reply]
All right, I just implemented it in #56.

**Reply by omichel** [reply]
It is now merged (and hence in production). Please update your local copy of the Nao proto with these changes: 06027aa9d8b0c280c46ce81d207739012771f7ea.
